### PR TITLE
Update on the paste layers

### PR DESCRIPTION
**Pull request by milg15**
I can delete the comments later 😅

### CODE REVIEW

File: packages/noya-state/src/selectors/layerSelectors.ts [hunk 1]
```diff
@@ -501,3 +501,15 @@ export function getConstrainedScaling(
       true)
   );
 }
+
+// Check if layer is visible in the page
+export function isLayerVisible(layer: Sketch.AnyLayer, canvasFrame: Rect) {
```
Comment (dabbott) [inline packages/noya-state/src/selectors/layerSelectors.ts:506]: I think we may want to use `getLayersInRect`, which will also handle transforms

https://github.com/noya-app/noya/blob/e3c896a128675123ec46608740f38d5f98767367/packages/noya-state/src/selectors/geometrySelectors.ts#L443-L452

File: packages/noya-state/src/reducers/layerReducer.ts [hunk 2]
```diff
@@ -438,50 +439,89 @@ export function layerReducer(
...
+          ) {
+            // If the layer is on the current page and is visible
             addSiblingLayer(draftPage, indexPaths[0], newLayer);
-          } else if (!Layers.isPageLayer(selectedLayer)) {
-            addSiblingLayer(draftPage, selectedLayerIndexPath, newLayer);
+          } else if (!Layers.isPageLayer(parentLayer)) {
+            // If the selected layer is a parent layer but not a page layer
+            newLayer.frame = {
+              ...newLayer.frame,
+              x: parentLayer.frame.width / 2 - layerFrame.width / 2,
```
Comment (dabbott) [inline packages/noya-state/src/reducers/layerReducer.ts:487]: To get a layer midpoint, we can use 

```ts
const parentBounds = getBounds(parentLayer.frame)
parentBounds.midX
parentBounds.midY
```

Comment (dabbott) [reply packages/noya-state/src/reducers/layerReducer.ts:487]: And maybe a function that sets the midpoint of a frame would help?

Untested:

```ts
// noya-geometry rect.ts
export function moveRectCentered<T extends Rect>(rect: T, point: Point): T {
  const bounds = createBounds(rect);

  const delta: Point = {
    x: bounds.midX - point.x,
    y: bounds.midY - point.y,
  };

  return {
    ...rect,
    x: rect.x + delta.x,
    y: rect.y + delta.y,
  };
}
```

**Comment by dabbott** [comment]
Looks good! Leave the comments, they're helpful 😁. Maybe just clean em up a little

It'll also be helpful to see how the tests are changed by these updates

**Review by dabbott — CHANGES_REQUESTED** [review]

File: packages/noya-state/src/reducers/layerReducer.ts (hunk 2 above)
Comment (milg15) [reply packages/noya-state/src/reducers/layerReducer.ts:487]: The parents midx and y are in relationship to the canvas not internal x and y so it doesn't work

**Review by dabbott — APPROVED** [review]
The new behavior looks great! 

I merged with main and refactored the frame/bounds stuff a little, but thanks to the tests I think I didn't break anything 😆

**Pull request merged**